src/qibo/transpiler/router.py

```diff
@@ -793,14 +833,31 @@ def _update_front_layer(self):
         """
         self._front_layer = self._get_dag_layer(0)
 
-    def _get_dag_layer(self, n_layer):
-        """Return the :math:`n`-topological layer of the dag."""
+    def _get_dag_layer(self, n_layer, qubits=False):
+        """Return the :math:`n`-topological layer of the dag.
+        If ``qubits=True``, return the target qubits of the blocks in the layer.
+        Otherwise, return the block numbers.
+        """
+
+        # 3# depend on the 'qubits' flag, return the block number or target qubits
+        # 3# return target qubits -> to avoid using get_physical_qubits(block_num)
+        if qubits:
+            layer_qubits = []
+            nodes = self._dag.nodes(data=True)
+            for node in nodes:
+                if node[1]["layer"] == n_layer:
+                    # return target qubits
+                    layer_qubits.append(node[1]["qubits"])
+            return layer_qubits
```

Comment:
```suggestion
            return [
                node[1]["qubits"]
                for node in self._dag.nodes(data=True) if node[1]["layer"] == n_layer 
            ]
```
this spares a couple of memory assignements